### ISSUE TITLE
Potential bug in mean centering in PCA?

### ISSUE THREAD

**Issue by Reason239** [IBM/activation-steering#6]
Hi!
I'm confused by the mean centering for PCA:

https://github.com/IBM/activation-steering/blob/59f24bf73817cfe3a836697c956359437aed0c8a/activation_steering/steering_vector.py#L189-L204

The text of the paper seems to imply that you subtract the mean *vector*:
<img width="968" alt="Image" src="https://github.com/user-attachments/assets/1c1f3dde-48c3-4087-8b5b-01b5768db330" />

as the notation $$h_1^+ - \mu_l$$ doesn't make sense if $$\mu_l$$ is a matrix and not a vector.

But in the text of the paper and in the code you define this mean $$\mu_l$$ as the half-sum of two *matrices*.

Shouldn't it be something like 

`center = ((h[::2] + h[1::2]) / 2).mean(axis=-1)`

Or just

`center = h.mean(axis=-1)`

?

As implemented now, when `method == "pca_center"`, `center` is a matrix, and the resulting `train` will have rows alternating between $$(h_i^+ - h_i^-)/2$$ and $$(h_i^- - h_i^+)/2$$. Which is equivalent to constructing `train` with `method == "pca_diff"` and "alternatively" concatenating `train/2` with `-train/2`. Is this intended?

I also didn't quite understand how the centering is motivated: is there is an overall offset in the data, then won't it be cancelled when taking the difference of vectors for `train` when `method == "pca_diff"` anyway? 

Moreover, the sklearn PCA implementation seems to apply feature centering to the input. So `method == "pca_diff"` implementation also applies centering? Isn't that a bit weird? Because if `h[::2] - h[1::2]` is offset from the origin, it seems that the steering vector is actually this offset, but the centering will destroy it?

Illustration of the `train` matrices:

<img width="624" alt="Image" src="https://github.com/user-attachments/assets/d23c6f02-2636-4fa8-949a-b638a115b9a2" />

**Comment by brucewlee** [comment]
Hi @Reason239, thank you for this detailed analysis! This seems to be an implementation issue for the public version. 

You're right that the current implementation creates pairwise centers (matrices) rather than computing a global mean vector, which deviates from standard mean-centering practice and our intended design.

The implementation should be:

```
if method == "pca_center":
    mu_l = h.mean(axis=0, keepdims=True) 
    train = h - mu_l 
```

Thank you again for this thorough analysis. It helps improve the reproducibility and correctness of our work. We'll make sure to credit you in the fix commit.

**Comment by brucewlee** [comment]
Hi @Reason239, thank you for this detailed analysis! This seems to be an implementation issue for the public version. 

The current code creates pairwise centers rather than implementing true mean-centering as described in the [paper](https://arxiv.org/abs/2312.03813) that inspired this approach.

For the pca_center method, we intended to:

1. First, removing the global mean 

2. Then, applying PCA to find the direction of maximum variance

The correct implementation should be:

```
if method == "pca_center":
    mu_global = h.mean(axis=0, keepdims=True) 
    train = h - mu_global  
```

Thank you for catching this!

**Comment by brucewlee** [comment]
We pushed a new version! It should work as expected now.
